packages/playwright-core/src/server/injected/recorder/recorder.ts

```diff
@@ -1471,18 +1471,29 @@ function removeEventListeners(listeners: (() => void)[]) {
   listeners.splice(0, listeners.length);
 }
 
+function frameDepth() {
+  let depth = 0;
+  // eslint-disable-next-line no-restricted-globals
+  let w: Window = window;
+  while (w.parent !== w) {
+    w = w.parent;
+    depth++;
+  }
+  return depth;
+}
+
 function querySelector(injectedScript: InjectedScript, selector: string, ownerDocument: Document): { selector: string, elements: Element[] } {
+  const empty = { selector, elements: [] };
   try {
-    const parsedSelector = injectedScript.parseSelector(selector);
-    return {
-      selector,
-      elements: injectedScript.querySelectorAll(parsedSelector, ownerDocument)
-    };
+    // selector starts from the root frame, we need to figure out if this frame is the right one
+    // as a cheap heuristic, we check if it targets the right depth
+    const parts = injectedScript.splitSelectorByFrame(selector);
+    const selectorTargetsFrameDepth = parts.length - 1 === frameDepth();
```

Comment:
I now realize this would regress quick locator checks where I type `getByRole('button')` in inspector and see all buttons on the page, including in iframes. Perhaps we should only check that iframe depth is more or equal than the selector depth?

Reply:
sounds good!